Sources/ClientRuntime/Networking/Http/URLSession/URLSessionHTTPClient.swift

```diff
@@ -0,0 +1,313 @@
+import class Foundation.URLSession
+import class Foundation.URLSessionConfiguration
+import class Foundation.URLSessionTask
+import class Foundation.URLSessionDataTask
+import protocol Foundation.URLSessionDataDelegate
+import AwsCommonRuntimeKit
+
+/// A client that can be used to make requests to AWS services using `Foundation`'s `URLSession` HTTP client.
+///
+/// This client is usable on all Swift platforms that support both the `URLSession` library and Objective-C interoperability features
+/// (these are generally the Apple platforms.)
+///
+/// Use of this client is recommended on all Apple platforms, and is required on Apple Watch ( see
+/// [TN3135: Low-level networking on watchOS](https://developer.apple.com/documentation/technotes/tn3135-low-level-networking-on-watchos)
+/// for details about allowable modes of networking on the Apple Watch platform.)
+///
+/// On Linux platforms, we recommend using the CRT-based HTTP client for its configurability and performance.
+public final class URLSessionHTTPClient: HTTPClient {
+
+    /// Holds a connection's associated resources from the time the connection is executed to when it completes.
+    private final class Connection {
+
+        /// The `FoundationStreamBridge` for the request body, if any.
+        ///
+        /// This reference is stored with the connection so that it may be closed (and its resources disposed of)
+        /// if the connection fails.
+        let streamBridge: FoundationStreamBridge?
+
+        /// The continuation for the asynchronous call that was made to initiate this request.
+        ///
+        /// Once the initial response is received, the continuation is called, and is subsequently set to `nil` so its
+        /// resources may be deallocated.
+        var continuation: CheckedContinuation<HttpResponse, Error>?
+
+        /// Any error received during a delegate callback for this request.
+        ///
+        /// The stored error is thrown back to the caller once the URLSessionDelegate receives
+        /// `urlSession(_:task:didCompleteWithError)` for this connection.
+        var error: Error?
+
+        /// A response stream that streams the response back to the caller.  Data is buffered in-memory until read by the caller.
+        let responseStream = BufferedStream()
+
+        /// Creates a new connection object
+        /// - Parameters:
+        ///   - streamBridge: The `FoundationStreamBridge` for the connection.
+        ///   - continuation: The continuation object for the `execute(request:)` call that initiated this connection.
+        init(streamBridge: FoundationStreamBridge?, continuation: CheckedContinuation<HttpResponse, Error>) {
+            self.streamBridge = streamBridge
+            self.continuation = continuation
+        }
+    }
+
+    /// Provides thread-safe associative storage of `Connection`s keyed by their `URLSessionDataTask`.
+    private final class Storage: @unchecked Sendable {
```

Comment:
I did not use `actor` for this class because it must be accessible from code that is async and also from code that is non-async (i.e. the `URLSessionDataTask` delegate callbacks.)

`NSRecursiveLock` conforms to `Sendable` so it is safe to use in both contexts.